### ISSUE TITLE
Portability of parquet/processor_encode

### ISSUE THREAD

**Issue by jshlbrd** [benthosdev/benthos#1851]
I was looking for a JSON -> Parquet converter and came across [parquet/processor_encode](https://github.com/benthosdev/benthos/blob/main/internal/impl/parquet/processor_encode.go), it works pretty well when ported outside of the project but I ran into an edge case I'm curious if anyone can figure out. Repeated complex objects (like `{"A":[{"B":1.0,"C":2.0},{"B":3.0,"C":4.0}]}`) cause panics in the go-parquet package: `panic: runtime error: index out of range [2] with length 2`. 

The error always reports the same number of elements that are in the repeated field. However, it works fine when the array contains one object (`{"A":[{"B":1.0},{"B":3.0}]}`) and more confusingly, it works fine when run inside Benthos. Below is an example program to test with, note that the internal Benthos query package is commented out since it isn't public:

```go

package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"

	"github.com/segmentio/parquet-go"
)

type Schema struct {
	A []struct {
		B float64
		C float64
	}
}

func main() {
	b := []byte(`{"A":[{"B":1.0,"C":2.0},{"B":3.0,"C":4.0}]}`)

	var m map[string]any
	if err := json.Unmarshal(b, &m); err != nil {
		fmt.Println(err)
	}

	f, err := os.Create("/tmp/test.parquet")
	if err != nil {
		panic(err)
	}

	ps := parquet.SchemaOf(new(Schema))
	w := parquet.NewGenericWriter[any](f, ps)
	defer w.Close()

	row, err := new(inserterConfig).toPQValuesGroup(ps.Fields(), m, 0, 0)
	if err != nil {
		panic(err)
	}

	if _, err := w.WriteRows([]parquet.Row{row}); err != nil {
		panic(err)
	}
}

//------------------------------------------------------------------------------

type inserterConfig struct {
	colIndex      int
	firstRepOf    *int
	parentMissing bool
}

func (c *inserterConfig) toPQValue(f parquet.Field, data any, defLevel, repLevel int) (parquet.Row, error) {
	if f.Repeated() {
		repLevel++

		var arr []any
		if data != nil {
			var isArray bool
			if arr, isArray = data.([]any); !isArray {
				return nil, fmt.Errorf("expected array, got %T", data)
			}
		}

		if len(arr) > 0 {
			defLevel++
		}

		return c.toPQValuesRepeated(f, arr, defLevel, repLevel)
	}

	if data == nil && !f.Optional() && !c.parentMissing {
		return nil, errors.New("missing and non-optional")
	}

	return c.toPQValueNotRepeated(f, data, defLevel, repLevel)
}

func (c *inserterConfig) toPQValueNotRepeated(f parquet.Field, data any, defLevel, repLevel int) (parquet.Row, error) {
	if len(f.Fields()) > 0 {
		var obj map[string]any
		if data != nil {
			var isObj bool
			if obj, isObj = data.(map[string]any); !isObj {
				return nil, fmt.Errorf("expected object, got %T", data)
			}
		}

		cCopy := *c
		if f.Optional() {
			if obj != nil {
				defLevel++
			}
			cCopy.parentMissing = true
		}

		res, err := cCopy.toPQValuesGroup(f.Fields(), obj, defLevel, repLevel)
		if err != nil {
			return nil, err
		}
		c.colIndex = cCopy.colIndex
		c.firstRepOf = cCopy.firstRepOf
		return res, err
	}

	var leafValue parquet.Value
	if data == nil {
		leafValue = parquet.ValueOf(nil)
	} else {
		if f.Optional() {
			defLevel++
		}

		switch f.Type().Kind() {
		// case parquet.Boolean:
		// 	b, err := query.IGetBool(data)
		// 	if err != nil {
		// 		return nil, err
		// 	}
		// 	leafValue = parquet.ValueOf(b)
		// case parquet.Int32:
		// 	iv, err := query.IGetInt(data)
		// 	if err != nil {
		// 		return nil, err
		// 	}
		// 	leafValue = parquet.ValueOf(int32(iv))
		// case parquet.Int64:
		// 	iv, err := query.IGetInt(data)
		// 	if err != nil {
		// 		return nil, err
		// 	}
		// 	leafValue = parquet.ValueOf(iv)
		// case parquet.Int96:
		// 	return nil, errors.New("columns of type Int96 are not currently supported")
		// case parquet.Float:
		// 	fv, err := query.IGetNumber(data)
		// 	if err != nil {
		// 		return nil, err
		// 	}
		// 	leafValue = parquet.ValueOf(float32(fv))
		// case parquet.Double:
		// 	fv, err := query.IGetNumber(data)
		// 	if err != nil {
		// 		return nil, err
		// 	}
		// 	leafValue = parquet.ValueOf(fv)
		// case parquet.ByteArray:
		// 	bv, err := query.IGetBytes(data)
		// 	if err != nil {
		// 		return nil, err
		// 	}
		// 	leafValue = parquet.ValueOf(bv)
		default:
			leafValue = parquet.ValueOf(data)
		}
	}

	if c.firstRepOf != nil {
		repLevel = *c.firstRepOf - 1
	}
	leafValue = leafValue.Level(repLevel, defLevel, c.colIndex)
	c.colIndex++

	return parquet.Row{leafValue}, nil
}

func (c *inserterConfig) toPQValuesRepeated(field parquet.Field, data []any, defLevel, repLevel int) (parquet.Row, error) {
	if len(data) == 0 {
		if c.firstRepOf == nil {
			c.firstRepOf = &repLevel
		}

		v, err := c.toPQValueNotRepeated(field, nil, defLevel, repLevel)
		if err != nil {
			return nil, err
		}
		return v, nil
	}

	endColIndex := c.colIndex

	var row parquet.Row
	for i, e := range data {
		// Prevent column index from being incremented on each iteration
		cCopy := *c
		if i == 0 {
			if cCopy.firstRepOf == nil {
				cCopy.firstRepOf = &repLevel
			}
		} else {
			cCopy.firstRepOf = nil
		}

		v, err := cCopy.toPQValueNotRepeated(field, e, defLevel, repLevel)
		if err != nil {
			return nil, err
		}
		row = append(row, v...)

		// Save the highest seen column index
		if cCopy.colIndex > endColIndex {
			endColIndex = cCopy.colIndex
		}
	}
	c.colIndex = endColIndex
	return row, nil
}

// https://www.waitingforcode.com/apache-parquet/nested-data-representation-parquet/read
// https://stackoverflow.com/questions/43568132/dremel-repetition-and-definition-level
// https://blog.twitter.com/engineering/en_us/a/2013/dremel-made-simple-with-parquet
func (c *inserterConfig) toPQValuesGroup(fields []parquet.Field, data map[string]any, defLevel, repLevel int) (row parquet.Row, err error) {
	for _, f := range fields {
		v, err := c.toPQValue(f, data[f.Name()], defLevel, repLevel)
		if err != nil {
			return nil, fmt.Errorf("field %v: %w", f.Name(), err)
		}
		row = append(row, v...)
	}
	return
}
```

**Comment by mihaitodor** [comment]
Hey @jshlbrd, I'm not really sure what you're trying to achieve, but I think the issue comes from using `parquet.SchemaOf()` instead of constructing the schema manually like Benthos does [here](https://github.com/benthosdev/benthos/blob/1e84116d40ca6830ca7cf7b076259e61bdad568b/internal/impl/parquet/processor_encode.go#L105). While `parquet.SchemaOf()` emits a similar object, it's not really the same thing, and the rest of the code breaks because of that. I managed to get it to emit something that's a bit closer to what you need by doing the following:

```go
type Schema struct {
	A []struct {
		B float64
		C float64
	} `parquet:"A,list"`
}
```
But you'll still have to modify `toPQValuesGroup` to do the iteration differently and that requires quite a bit of extra investigation. Consider debugging both Benthos and your code in parallel. I used the following config:

```yaml
input:
  generate:
    mapping: |
      root = {"A":[{"B":1.0,"C":2.0},{"B":3.0,"C":4.0}]}
    count: 1
    interval: 0s
  processors:
    - parquet_encode:
        schema:
          - name: A
            repeated: true
            fields:
              - name: B
                type: FLOAT
              - name: C
                type: FLOAT
    - log:
        message: |
          Error: ${! error() }

output:
  stdout: {}
```

But, since it's not clear what you're trying to achieve, I'm not sure why this is needed. You might want to ask the `github.com/segmentio/parquet-go` maintainers for some guidance.

**Comment by jshlbrd** [comment]
@mihaitodor thanks for the response! The goal is to find alternative solutions to writing JSON to Parquet, ideally as a package. This code isn't exported, which is why I asked about the portability of it. I've come across https://github.com/xitongsys/parquet-go as well.

By the way, it appears that the issue is caused by the segmentio/parquet-go package -- I had the most recent version installed via go.mod, when I ported it back to the version currently used by Benthos the example program runs fine. I'll close this out since the question is answered, but you may run into issues upgrading that package in the future.

**Comment by mihaitodor** [comment]
I see, thank you @jshlbrd! That makes sense now... I raised #1855 to keep track of this. Initially, I thought it's only a breaking change in `parquet.SchemaOf()`, which Benthos isn't calling, but it looks like there's more to it, unfortunately :(

**Comment by mihaitodor** [comment]
PS: You might want to have a look at https://pkg.go.dev/github.com/apache/arrow/go/v9/parquet too. It's also on my list to investigate sometime... I had a chat with its author, Matt Topol, on the Gophers slack and he shared the following details about its performance as well as [this presentation](https://www.youtube.com/watch?v=ctK9MxDJd2Q&list=PLU2OcwpQkYCz6Umb4q-x82yehGekehQP7&index=3).

> We performed an experiment using Arrow's Parquet writer. In the experiment, we wrote 25 million rows to the Parquet file, each with 10 fields of type int32. This resulted in a Parquet file of about 1 GB in size, with just one row group. We allocated ~100 MB of data in-memory at a time before writing to the Parquet writer, and then freed this data by letting it go out of scope and forcing garbage collection.

> Memory used (total)
> This graph shows the total memory used in the Arrow Parquet Writer experiment. The peak data points were taken right after allocating 100 MB of input data (in column form) for the writer. As you can see, the Arrow writer appears to hold onto very little data. The trough data points are about 5 MB.
![image](https://user-images.githubusercontent.com/788216/232664169-1c79a303-2b8d-4f0b-9b21-9c0d0b622765.png)

**Comment by jshlbrd** [comment]
> I see, thank you @jshlbrd! That makes sense now... I raised #1855 to keep track of this. Initially, I thought it's only a breaking change in `parquet.SchemaOf()`, which Benthos isn't calling, but it looks like there's more to it, unfortunately :(

No problem!

> the parquet-specific code in Benthos isn't exported because it's just an implementation detail and it's not meant to be a generic utility library that others should rely on. Ideally, the `github.com/segmentio/parquet-go` authors should have a look at how Benthos uses this library and consider exposing some more API functions to make it easier to use.

That is my thought as well. Not sure if the Benthos team has the appetite to submit a PR for this on the segmentio library, but if they accepted the addition then maybe it would reduce the likelihood of breaking changes.
